R/step-join.R

```diff
@@ -1,26 +1,42 @@
 step_join <- function(x, y, on, style, suffix = c(".x", ".y")) {
   stopifnot(is_step(x))
   stopifnot(is_step(y))
-  stopifnot(is.character(on))
+  stopifnot(is.null(on) || is.character(on))
+  style <- match.arg(style, c("inner", "full", "right", "left", "semi", "anti"))
 
-  style <- match.arg(style, c("inner", "full", "left", "semi", "anti"))
-  if (style %in% c("semi", "anti")) {
-    vars <- x$vars
-  } else {
-    vars <- join_vars(x$vars, y$vars, on, suffix)
-  }
+  on <- dplyr::common_by(on, x, y)
 
-  new_step(
-    parent = x,
+  vars_out_dt <- dt_join_vars(x$vars, y$vars, on$x, on$y, suffix = suffix, style = style)
+  colorder <- dt_join_colorder(x$vars, y$vars, on$x, on$y, style)
+
+  # TODO suppress warning in merge
+  # "column names ... are duplicated in the result
+  out <- new_step(
+    parent = if (style == "left") y else x,
     implicit_copy = TRUE,
-    parent2 = y,
-    vars = vars,
-    on = on,
-    suffix = suffix,
+    parent2 = if (style == "left") x else y,
+    vars = vars_out_dt,
+    on = if (style %in% c("left", "full")) on else list(x = on$y, y = on$x),
     style = style,
     locals = utils::modifyList(x$locals, y$locals),
     class = "dtplyr_step_join"
   )
+
+  if (style %in% c("anti", "semi")) {
+    return(out)
+  }
+
+  out <- step_colorder(out, colorder)
```

Comment:
* `data.table::merge()` produces the order `join-vars, x-vars, y-vars`. `full_join()` produces the order `x-vars, y-vars`
* The left join in data.table is `y[x]` (instead of `x[y]`), so the order is `y-vars, x-vars` instead of `x-vars, y-vars` as in `left_join()`